fix: collect metrics on scrape, not timeout

sam-github opened the pull request:
Only the event loop "lag" is still async, see the in-src notes.
    
Fixes: https://github.com/siimon/prom-client/issues/180

---

This is a WIP, it lacks tests (though it passes the existing ones).

Open questions:
- should I rearrange the commits? There are some unrelated ones in the PR, I could pull them into a seperate PR set if wanted. I figured out a way to do this that doesn't change the lib/metrics interface for the defaults, for example, so the metrics changes are mostly cleanup and could be done with or without the removal of the timeout.
- are the lib/metrics/ used directly by anyone? I'd like to remove the do-nothing closures for many of them that are called on every collect but are a no-op.
- do we need to support reconfiguration of the default metrics on the ***same*** registry? I can do it, its just a bit more state that needs to be removed now, and some notes in source suggested it wasn't wanted. Then again, not doing it might break some users. Opinions?

Also, my typescript is close to non-existent, pay close attention to my attempt to update the typescript definitions.

SimenB commented on index.d.ts:
```diff
@@ -663,8 +663,8 @@ export function collectDefaultMetrics(
- * @param timeout The interval how often the default metrics should be probed
- * @deprecated A number to defaultMetrics is deprecated, please use \`collectDefaultMetrics({ timeout: ${timeout} })\`.
+ * @param timeout Deprecated, ignored if present.
+ * @deprecated Deprecated, ignored if present.
```
can we just remove the entire argument? next release is a major anyways

SimenB replied:
we can do so (plus other deprecated stuff) separately, perhaps?

sam-github replied:
how about do it now, it makes any subsequent changes easier if the deprecated code isn't there: https://github.com/siimon/prom-client/pull/330

SimenB replied:
Sure, whatever's easiest!

siimon replied:
Hell yeah! I'm all for removing all deprecated stuffs.. been there for so long so it should be ok to remove those things now

sam-github replied:
removed in #330 and I rebased this WIP on top of #330

sam-github commented:
No hurry, enjoy your weekend. I'm willing to remove code for sure, I'll get rid of the timeout relics.

SimenB commented:
>are the lib/metrics/ used directly by anyone?

it's not supported, so I'm not afraid of breaking people's usage if it. And since next release is a major, I think it's fine.

>do we need to support reconfiguration of the default metrics on the same registry?

I don't think so - what's the use case for it?

SimenB commented on lib/metrics/processHandles.js:
```diff
@@ -28,19 +28,11 @@ module.exports = (registry, config = {}) => {
-	return updater;
+	return () => {
+		const handles = process._getActiveHandles();
+		updateMetrics(gauge, aggregateByObjectName(handles), false);
```
do we need this `updateMetrics` helper now?

sam-github replied:
Yes, its shared code, used by lib/metrics/processRequests.js and lib/metrics/processHandles.js.

I think I can remove the second argument to aggregateByObjectName() though, its always passed `false`.

sam-github replied:
done

sam-github commented:
>> do we need to support reconfiguration of the default metrics on the same registry?
> I don't think so - what's the use case for it?

No idea, but existing code supports it:
- https://github.com/siimon/prom-client/blob/9b4ef105cbc914e0ca50d9fbca9113deeb3db663/lib/defaultMetrics.js#L36
- https://github.com/siimon/prom-client/blob/9b4ef105cbc914e0ca50d9fbca9113deeb3db663/lib/defaultMetrics.js#L61-L66

I can reproduce the master behaviour in this PR, but if the comment is correct, then its OK to just delete the support. Thought I'd ask, though, rather than try to sneak it into a larger change. Full disclosure and all. :-)

SimenB commented:
It seems I added that comment 3 years ago - I have no recollection of it 😅 

But no, I think it makes sense to throw if called again. IIRC it should do so if we just remove the guard, as that's the default registry behavior

siimon commented:
> > are the lib/metrics/ used directly by anyone?
> 
> it's not supported, so I'm not afraid of breaking people's usage if it. And since next release is a major, I think it's fine.
> 
> > do we need to support reconfiguration of the default metrics on the same registry?
> 
> I don't think so - what's the use case for it?

I agree, in my world the public interface to this library is what is in /index.js

siimon commented:
Regarding the reconfiguration I don't either see the use case for it. I think it could be removed 

> It seems I added that comment 3 years ago - I have no recollection of it 😅
> 
> But no, I think it makes sense to throw if called again. IIRC it should do so if we just remove the guard, as that's the default registry behavior

I agree here aswell. The only potential problem I could see is in tests where you run it multiple times. But that should probably be handled when you write the tests with a proper cleanup.

siimon commented:
I think it looks great, really appreciate your time working on this!

sam-github commented:
> The only potential problem I could see is in tests where you run it multiple times. But that should probably be handled when you write the tests with a proper cleanup.

Or write the test to use a non-default registry, the throw that is in there will trigger if the default metrics are added multiple times _to the same registry_. They can be added to a number of different registries.

sam-github commented:
Added a test for https://github.com/siimon/prom-client/pull/329#issuecomment-583175871

sam-github commented:
I addressed the comments, I guess its ready for review again.

It does contain #330, maybe that should land first, then I'll rebase this. Or this could land and close #330. Whatever you all prefer.

siimon commented:
@sam-github the changes LGTM! Just got a conflict after the latest merge, so if you could fix that and update the PR and I'll merge right away!

SimenB commented on CHANGELOG.md:
```diff
@@ -10,6 +10,8 @@ project adheres to [Semantic Versioning](http://semver.org/).
+- Dropped the previously deprecated support for positional parameters in
+  constructors, only the config object forms remain.
 
 ### Changed
```
should mention the change in behaviour here

SimenB commented on index.d.ts:
```diff
@@ -662,14 +606,6 @@ export function collectDefaultMetrics(
 	config?: DefaultMetricsCollectorConfiguration
 ): ReturnType<typeof setInterval>;
```
this return type is wrong now, right? should be `: void`

SimenB commented on lib/defaultMetrics.js:
```diff
@@ -31,57 +33,54 @@ const metrics = {
+	defaultMetricCollector();
 
-	return existingInterval;
+	return 0; // Backwards compat, used to be a timeout in milliseconds.
```
no need for this I think, it's landing in a major anyways. Just returning nothing seems fine

SimenB commented on lib/defaultMetrics.js:
```diff
@@ -31,57 +33,54 @@ const metrics = {
-// We might want to consider not supporting running the default metrics function more than once
-let init = true;
-
 module.exports = function startDefaultMetrics(config) {
```
```suggestion
module.exports = function collectDefaultMetrics(config) {
```

maybe? The name makes no real difference

SimenB commented on lib/metrics/eventLoopLag.js:
```diff
@@ -1,5 +1,22 @@
+// somewhat compatible with the current code.
+// - Might be possible to report perf_hook's histogram as a summary, using its
+// percentile() function. Summaries are less than ideal with prometheus but at
+// least the data is available in 12.x+. See #309
```
```suggestion
// least the data is available in 12.x+. See https://github.com/siimon/prom-client/issues/309
```

Just easier for people if they come over this in `node_modules` and want to take a look.

👏 for detailed comment, btw!

SimenB commented on lib/metrics/processOpenFileDescriptors.js:
```diff
@@ -20,14 +20,12 @@ module.exports = (registry, config = {}) => {
-				return;
-			}
-
-			// Minus 1, as this invocation created one
```
can we keep this comment?

sam-github commented:
@SimenB Thanks for the careful review, I made all the suggested changes, PTAL.
